### ISSUE TITLE
The cider score gets smaller than pretrained model

### ISSUE THREAD

**Issue by wjb123** [ruotianluo/self-critical.pytorch#14]
Hi, i am trying to reproduce your code in tensorflow. I almost write my code as you do, but find the cider score is getting smaller than before. At the same time, the sampling sentence is also getting shorter. Can you give me some advice?

**Comment by ruotianluo** [comment]
After hoe many epoch do you start self critical training.

**Comment by gonnafree** [comment]
Hi Luo, I also got the same problem. I trained on video captioning dataset, so your current setting does not apply. I aware that you mention not to start SCST training too late; however, so far I haven't found a suitable time to start SCST yet. The batch rewards are mostly negative (which means the greedy generated sentences are generally better than the sampling ones in terms of Cider-D). So the average reward and the loss value are often negative. Did you experience the same problem? Is this normal to have negative rewards and loss values with SCST training?

**Comment by ruotianluo** [comment]
Yes, it's exactly the case. Try to find the sweet spot. I should say it cant be too early or too late.

**Comment by gonnafree** [comment]
Thanks. I still struggle to search for this sweet spot. Did you use Scheduled sampling to pre-train the XE model?

**Comment by ruotianluo** [comment]
I did.

**Comment by wjb123** [comment]
I first pretrain a XE model，and then use it to sample sentences as your code, but find the cider score does not improve. I don't use Scheduled sampling, is it key to self-critical training?

**Comment by ruotianluo** [comment]
I dont think scheduled sampling is the key, but it can improve the result according to the authors.

**Comment by wjb123** [comment]
So why does my self-critical model get smaller cider value (on the val set) than pretrained model during the training stage ?

**Comment by ruotianluo** [comment]
when did you start self critical training. (I actually never tried without scheduled sampling with my code, I don't guarantee if it will work or not. )

**Comment by wjb123** [comment]
I start self-critical training from the first step, as the pretained XE model looks good.

**Comment by ruotianluo** [comment]
I meant how many epochs did you pretrain

**Comment by wjb123** [comment]
about 20 epochs

**Comment by ruotianluo** [comment]
Maybe it's the problem of the scheduled sampling. If you follow every step in my readme, the result is definitely reproducible.

**Comment by wjb123** [comment]
As my pretained model converges in the early epoches (6 to 10) and I save my best model according to the bleu4 value during training. Does this affect the later self-critical training ?

**Comment by ruotianluo** [comment]
Are you training on different data set?

**Comment by ruotianluo** [comment]
And which model are you using?

**Comment by wjb123** [comment]
I use fc-attention model, which is trained on the coco dataset.

**Comment by ruotianluo** [comment]
what is fc attention??

**Comment by wjb123** [comment]
the attention model from https://github.com/kelvinxu/arctic-captions.

**Comment by ruotianluo** [comment]
I see. I suggest you not to select the best model of pretraining . I'm now actually treating it as a two stage training. Just keep training from the latest checkpoint. The epoch starting self critical training matters a lot.

**Comment by wjb123** [comment]
ok, I will try as you suggest.

**Comment by qingchenwuhou** [comment]
hi @wjb123 , you said " ... but find the cider score does not improve. I don't use Scheduled sampling, is it key to self-critical training?", 
1. Do you mean that you use Schedule sampling in the first stage training and do not use in self-critical training, or you don't use Schedule sampling in both training stages?
2. Does the trouble of decreasing CIDEr score in second stage training be solved, and how you solve it?
Many Thanks!

**Comment by qingchenwuhou** [comment]
I think the mistake I make is that I didn't use the best model and corresponding infos.pkl as the initialization of SCST training, as detailed in paper:
" We evaluate at each epoch the model on the development set and select the model with best CIDEr score as an initialization for SCST training"
